tests/inference/qnn_test.py

```diff
@@ -475,41 +496,125 @@ def generate_resolvers():
+      bias_initializer = tf.keras.initializers.RandomUniform(
+          minval=min_val,
+          maxval=max_val,
+          seed=(self.tf_random_seed_alt + 2 * i + 1))
       expected_layer_list.append(
-          tf.keras.layers.Dense(units[i], activation=activations[i]))
-    expected_layer_list.append(tf.keras.layers.Dense(1))
+          tf.keras.layers.Dense(
+              units[i],
+              activation=activations[i],
+              kernel_initializer=kernel_initializer,
+              bias_initializer=bias_initializer))
+    kernel_initializer = tf.keras.initializers.RandomUniform(
+        minval=min_val, maxval=max_val, seed=(self.tf_random_seed_alt + 2 * i))
+    bias_initializer = tf.keras.initializers.RandomUniform(
+        minval=min_val,
+        maxval=max_val,
+        seed=(self.tf_random_seed_alt + 2 * i + 1))
+    expected_layer_list.append(
+        tf.keras.layers.Dense(
+            1,
+            kernel_initializer=kernel_initializer,
+            bias_initializer=bias_initializer))
     expected_layer_list.append(utils.Squeeze(-1))
     hamiltonian_energy = models.BitstringEnergy(bits, expected_layer_list)
+    hamiltonian_energy.build([None, self.num_bits])
+    # TODO(#209)
+    _ = hamiltonian_energy(tf.constant([[0] * self.num_bits], tf.int8))
     hamiltonian = models.Hamiltonian(hamiltonian_energy, hamiltonian_circuit)
 
     # Get expectations
     total_resolvers = generate_resolvers()
     qb_keys = [(q, f"measure_qubit_{i}") for i, q in enumerate(qubits)]
-    expected_expectations = []
+    raw_expectations = []
     for r in total_resolvers:
       samples_pd = cirq.Simulator().sample(
           total_circuit, repetitions=expectation_samples, params=r)
       samples = samples_pd[[x[1] for x in qb_keys]].to_numpy()
       current_energies = hamiltonian_energy(samples)
-      expected_expectations.append(tf.math.reduce_mean(current_energies))
+      raw_expectations.append(
+          tf.math.reduce_mean(current_energies, keepdims=True))
+    expected_expectations = tf.stack(raw_expectations)
     self.assertNotAllClose(
         expected_expectations,
         tf.zeros_like(expected_expectations),
         atol=self.not_zero_atol)
 
     # Compare
-    actual_expectations = actual_qnn.expectation(initial_states, hamiltonian)
+    expectation_wrapper = tf.function(actual_qnn.expectation)
+    actual_expectations = expectation_wrapper(initial_states, hamiltonian)
     self.assertAllClose(
         actual_expectations, expected_expectations, rtol=self.close_rtol)
+    self.assertAllEqual(
+        tf.shape(actual_expectations), [len(initial_states_list), 1])
+
+    # TODO(#205) function seems to be ripe for refactoring
+    def delta_expectations_func(k, var, delta):
+      """Calculate the expectation with kth entry of `var` perturbed."""
+      num_elts = tf.size(var)
+      old_value = var.read_value()
+      # Can no longer assume var is 1D, since energy uses matrix variables now
```

Comment:
can we use this adapted version of the fn for `test_expectation_cirq`?

Reply:
The code there is slightly different right now, but I think we can refactor this out eventually.   Hence #205

Reply:
I would bias towards not introducing more clutter and taking out another TODO loan. You're call

Reply:
The code is different thought, so n my opinion it would be more cluttered to do the refactoring now.  Properly it should be a test util function that is tested, then we can confidently use a generic version (similar code also shows up in `ebm_test.py`, agains slightly different).  In my opinion the proper generic version is a function that accepts the expectation value producing function as an argument, that we would `functools.partial` away in each test

Reply:
Makes sense - will be a high yield TODO for readability then IMO